trio/_subprocess.py

```diff
@@ -1,73 +1,273 @@
+                self.kill()
+                with _core.open_cancel_scope(shield=True):
+                    await self.wait()
+
+    async def wait(self):
+        """Block until the process exits.
+
+        Returns:
+          The exit status of the process (a nonnegative integer, with
+          zero usually indicating success). On UNIX systems, a process
+          that exits due to a signal will have its exit status reported
+          as the negative of that signal number, e.g., -11 for ``SIGSEGV``.
+        """
+        if self.poll() is None:
+            await wait_child_exiting(self._proc)
+            self._proc.wait()
+        else:
+            await _core.checkpoint()
+        return self.returncode
+
+    def poll(self):
+        """Forwards to :meth:`subprocess.Popen.poll`."""
+        return self._proc.poll()
+
+    def send_signal(self, sig):
+        """Forwards to :meth:`subprocess.Popen.send_signal`."""
+        self._proc.send_signal(sig)
+
+    def terminate(self):
+        """Forwards to :meth:`subprocess.Popen.terminate`."""
+        self._proc.terminate()
+
+    def kill(self):
+        """Forwards to :meth:`subprocess.Popen.kill`."""
+        self._proc.kill()
+
+
+async def run(
+    *popenargs, input=None, timeout=None, deadline=None, check=False, **kwargs
+):
+    """Like :func:`subprocess.run`, but async.
+
+    Unlike most Trio adaptations of standard library functions, this
+    one keeps the ``timeout`` parameter, so that it can provide you
+    with the process's partial output if it is killed due to a
+    timeout. It also adds ``deadline`` as an option if you prefer to
+    express your timeout absolutely. If you don't care about preserving
+    partial output on a timeout, you can of course also nest run()
+    inside a normal Trio cancel scope.
+
+    Returns:
+      A :class:`subprocess.CompletedProcess` instance describing the
+      return code and outputs.
+
+    Raises:
+      subprocess.TimeoutExpired: if the process is killed due to timeout
+          expiry
+      subprocess.CalledProcessError: if check=True is passed and the process
+          exits with a nonzero exit status
+      OSError: if an error is encountered starting or communicating with
+          the process
+
+    """
+    if input is not None:
+        if 'stdin' in kwargs:
+            raise ValueError('stdin and input arguments may not both be used')
+        kwargs['stdin'] = subprocess.PIPE
+
+    if timeout is not None and deadline is not None:
+        raise ValueError('timeout and deadline arguments may not both be used')
+
+    stdout_chunks = []
+    stderr_chunks = []
+
+    async with Process(*popenargs, **kwargs) as proc:
+
+        async def feed_input():
+            if input:
+                try:
+                    await proc.stdin.send_all(input)
+                except _core.BrokenResourceError:
+                    pass
+                except OSError as e:  # pragma: no cover
```

Comment:
Already had a test like that, which didn't trigger the issue. I dug into the bugs linked in stdlib subprocess and found (https://bugs.python.org/issue30418)

> The underlying Windows error is ERROR_NO_DATA. If we could know that for certain, then we could ignore it as a BrokenPipeError. Currently all we have to go on from _Py_write is the CRT's EINVAL errno value. In contrast, when creating an OSError from the Windows last error value, winerror_to_errno() maps ERROR_NO_DATA as EPIPE.

So, since we're calling the win32 API ourselves, we get an EPIPE and don't need the hack.